flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java

```diff
@@ -150,28 +166,48 @@ public String addressString() {
 
 	/**
 	 * Returns the fully-qualified domain name the TaskManager. If the name could not be
-	 * determined, the return value will be a textual representation of the TaskManager's IP address.
+	 * determined or {@link JobManagerOptions.RETRIEVE_TASK_MANAGER_HOSTNAME} is set to false,
+	 * the return value will be a textual representation of the TaskManager's IP address.
 	 * 
 	 * @return The fully-qualified domain name of the TaskManager.
 	 */
 	public String getFQDNHostname() {
+		// Lazily retrieve FQDN host name of this TaskManager
+		if (fqdnHostName == null) {
+			if (!retrieveHostName) {
+				fqdnHostName = inetAddress.getHostAddress();
+			} else {
+				fqdnHostName = getFqdnHostName(inetAddress);
+			}
+		}
+
 		return fqdnHostName;
 	}
 
 	/**
-	 * Gets the hostname of the TaskManager. The hostname derives from the fully qualified
-	 * domain name (FQDN, see {@link #getFQDNHostname()}):
+	 * Gets the hostname of the TaskManager.
+	 * If {@link JobManagerOptions.RETRIEVE_TASK_MANAGER_HOSTNAME} is true,
```

Comment:
again, this class is exposing an implementation detail of `JobMaster`.

(I'm a bit uncertain about this review item, maybe it is also okay. But I believe it is dangerous to mention this here as the implementation could change)